lib/puppet/functions/simplib/ip/family_hash.rb

```diff
@@ -0,0 +1,82 @@
+# Process an array of IP addresses and return them split by IP family and
+# include metadata and/or processed versions.
+#
+Puppet::Functions.create_function(:'simplib::ip::family_hash') do
+  # @param ip_addresses
+  #   The addresses to convert
+  #
+  # @return [Hash]
+  #   Converted Hash with the following format (YAML representation):
+  #
+  #   ```
+  #   # IPv4 Addresses
+  #   ipv4:
+  #     <Passed Address>:
+  #       address: <normalized address>
+  #       netmask:
+  #         ddq: <dotted quad notation netmask>
+  #         cidr: <CIDR netmask>
+  #   # IPv6 Addresses
+  #   ipv6:
+  #     <Passed Address>:
+  #       address: <normalized address>
+  #       netmask:
+  #         # DDQ is not valid for IPv6
+  #         ddq: nil
+  #         cidr: <CIDR netmask>
+  #   ```
+  #
+  dispatch :family_hash do
+    required_param 'Variant[
+      Simplib::IP,
+      Simplib::IP::V4::DDQ,
+      Simplib::IP::V4::CIDR,
+      Simplib::IP::V6::CIDR,
+      Array[Variant[
+        Simplib::IP,
+        Simplib::IP::V4::DDQ,
+        Simplib::IP::V4::CIDR,
+        Simplib::IP::V6::CIDR
+      ]]
+    ]', :ip_addresses
+  end
+
+  def family_hash(addresses)
+    results = {}
+
+    Array(addresses).uniq.each do |addr|
```

Comment:
This eliminates duplicate input addresses (undocumented feature), but doesn't eliminate entries for which their metadata is the same.  For example, 1.2.0.0/16 is the same as 1.2.3.4/16.  Should the effective duplicates be removed as well?

Reply:
Honestly, I'm not sure. I wanted to leave it in for now to see if it would be an issue in the future. I almost used the normalized address as the key which would also solve that issue. @lnemsick-simp which way do you think it should go?

Reply:
I think this should return the normalized result.  All duplicates removed.  If the code you are writing doesn't care (the only user of the function at this point), I would definitely make the change.

Reply:
Turns out that I can't do this because the 'normalized' address is the bare address without the netmask (and needs to be that way to prevent additional re-splitting work later on). Reversing the two would make usage too difficult.